hack/generator/pkg/astmodel/type_walker.go

```diff
@@ -0,0 +1,194 @@
+	}
+
+	def, ok := t.allTypes[it]
+	if !ok {
+		return nil, errors.Errorf("couldn't find type %q", it)
+	}
+
+	// Prevent loops by bypassing this type if it's currently being processed.
+	if _, ok := t.state.processing[def.Name()]; ok {
+		remove, err := t.ShouldRemoveCycle(def, updatedCtx)
+		if err != nil {
+			return nil, err
+		}
+		if remove {
+			return typeWalkerRemoveType, nil
+		} else {
+			// Preserve
+			return def.Name(), nil
+		}
+	}
+	t.state.processing[def.Name()] = struct{}{}
+
+	updatedType, err := this.Visit(def.Type(), updatedCtx)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error visiting type %q", def.Name())
+	}
+	updatedDef, err := t.AfterVisit(def, def.WithType(updatedType), updatedCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	delete(t.state.processing, def.Name())
+
+	err = t.state.result.AddWithEqualityCheck(updatedDef)
+	if err != nil {
+		return nil, err
+	}
+
+	return updatedDef.Name(), nil
+}
+
+func (t *TypeWalker) visitObjectType(this *TypeVisitor, it *ObjectType, ctx interface{}) (Type, error) {
+	result, err := t.originalVisitObjectType(this, it, ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	ot, ok := result.(*ObjectType)
+	if !ok {
+		panic(fmt.Sprintf("TypeWalker visitor VisitObjectType must return a *ObjectType, instead returned %T", result))
+	}
+
+	for _, prop := range ot.Properties() {
+		shouldRemove := isRemoveType(prop.PropertyType())
+		if shouldRemove {
+			ot = ot.WithoutProperty(prop.PropertyName())
+		}
+	}
+
+	return ot, nil
+}
+
+// Walk returns a Types collection constructed by applying the Visitor to each type in the graph of types reachable
+// from the provided TypeDefinition 'def'. Types are visited in a depth-first order. Cycles are not followed
+// (so each type in a cycle will be visited only once).
+func (t *TypeWalker) Walk(def TypeDefinition) (Types, error) {
+	t.state = typeWalkerState{
+		result:     make(Types),
+		processing: make(map[TypeName]struct{}),
+	}
+
+	// Visit our own name to start the walk.
+	// Initial ctx is nil -- MakeContext will get called and fabricate a context if needed
+	_, err := t.visitor.Visit(def.Name(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return t.state.result, nil
+}
+
+// IdentityAfterVisit is the default AfterVisit function for TypeWalker. It returns the TypeDefinition from Visit unmodified.
+func IdentityAfterVisit(_ TypeDefinition, updated TypeDefinition, _ interface{}) (TypeDefinition, error) {
+	return updated, nil
+}
+
+// IdentityMakeContext returns the context unmodified
+func IdentityMakeContext(_ TypeName, ctx interface{}) (interface{}, error) {
+	return ctx, nil
+}
+
+// IdentityShouldRemoveCycle is the default cycle removal behavior. It preserves all cycles unmodified.
+func IdentityShouldRemoveCycle(_ TypeDefinition, _ interface{}) (bool, error) {
+	return false, nil
+}
+
+func isRemoveType(t Type) bool {
```

Comment:
Perhaps there's a better name? Here's my stab at something more intention revealing:

```suggestion
func shouldBeRemoved(t Type) bool {
```

Reply:
I dropped the `Be` and just renamed it to `shouldRemove`.